---
title: "ACS2022MEL_scRNAseq_demo"
author: "Wenyan"
date: '2022-11-14'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### This is a RMarkdown file, which has separate sections for texts and R scripts. 
### It's a good tool to build analytical reports.

## Step R environment

### 1. Install R packages
```{r}
if (!require(Seurat)) install.packages("Seurat")
if (!require(tidyverse)) install.packages("tidyverse")
if (!require(patchwork)) install.packages("patchwork")
```

### 2. Load R packages
```{r}
library(Seurat)
library(tidyverse)
library(patchwork)
```

### 3. Working directory
```{r}
# check the current working directory
getwd()
```

```{r}
# use a path as working directory
setwd(dir = "/Users/wli/Documents/scRNAseq_Git_GitHub/ACS_MEL_2022") 

# try another path to test
# setwd(dir = "/Users/wli/Documents/") 
```

### or use the Rstudio "File" tab on the right to choose the 
### folder you want to set as working directory

### However, when you create a new project, the working directory will be the 
### folder that you choose to store your project files

### =================
### Note: 
In this working directory (or the folder you choose), you should have downloaded
1. "demo_ACS2022MEL.Rmd" file
2. 10x demo data (instruction to do so provided in the README.md file)
3. Rhapsody demo data  (instruction to do so provided in the README.md file)

And make sure you have installed packages:
1. Seurat
2. tidyverse
3. patchwork
### =================


## R basis
### 1. Numeric
```{r}
# numbers
x <- 3 # assign 3 to x

# print x out
x 
```

```{r}
# vector
y <- c(1, 3, 5) # assign 1, 3 and 5 to 5

# print y out
y 
```

```{r}
# concatenate x and y
z <- c(x, y) 

# print z out
z 
```

### 2. Strings
```{r}
a <- "I am a string"
a
```

```{r}
# string vectors
b <- c("oh", "my", "god")
b
```

```{r}
# concatenate strings
c <- c(a, b)
c
```

### 3. Data frame
```{r}
# dataframe
df <- data.frame(col1 = c(1:3)*10, 
                 col2 = y, 
                 col3 = b, 
                 row.names = c("row name 1", 
                               "row name 2", 
                               "row name 3"))
df
```

```{r}
# subset a column or more
df$col1

df[c("col1", "col3")]

```

```{r}
# subset based on criterion 
df[df$col1 > 10, ]

# but only want data from col3
df[df$col1 > 10, "col3"]
```

### 4. List
```{r}
# you can store different data types all in one R objects. The datasets do need to be
# the same size.
a_list <- list("xman", "superman", "spiderman", c(1:5))
a_list

```

```{r}
# or use existing objects
b_list <- list(df, x, a)
b_list

```

```{r}
# assign names to each list of b_list
names(b_list) <- c("df", "number", "string")

# extract elements from a list
b_list$number

```

### 5. logical operators
```{r}
# logical operators
saveFile <- TRUE # or just T
saveCSV <- F
```

## scRNAseq analysis

### 1. Read gene expression files to R
###    1a. BD Rhapsody platform
           Tell Rstudio where the files are
```{r}
# Data from Rhapsody
# The demo data selected was created from a WTA pipeline. Hence, use
# the file ends with RSEC_MolsPerCell.csv

# method 1 - manually add find the file path
rhap_file <- "Rhapsody PBMC WTA AbSeq SMK/Combined_BD-Demo-WTA-AbSeq-SMK_RSEC_MolsPerCell.csv" 

# method 2 - search file that contains this string "RSEC_MolsPerCell.csv"
rhap_file2 <- list.files("Rhapsody PBMC WTA AbSeq SMK", 
                         full.names = T)[grep("RSEC_MolsPerCell.csv", 
                                              list.files("Rhapsody PBMC WTA AbSeq SMK", 
                                                         full.names = T))]
```

Then you can load the file(s).
Here, a csv file will be uploaded.
```{r}
# load file as a matrix
mat <- read.table(file = rhap_file, # path of the file
                  sep = ",",        # column separator
                  header = T)       # use first row as column name

# check mat 
mat[1:10, 1:10]

```

Once the data is loaded to the environment, you can interact with it
```{r}
# 1. use Cell_Index column as row names and remove this column
# 2. transpose matrix

# this operator "%>%" works as a pipe.
# Object (calculation result) on the left side of "%>%" passes to the right hand side of the operator.

mat <- mat %>%    
       tibble::column_to_rownames(var = "Cell_Index") %>% 
       t()

mat[1:10, 1:10]
```

Inspect the matrix
```{r}
# Protein and RNA reads are mixed in the matrix. We need to separate them.
# Protein oligos end with pAbO. It can be used as a protein selector.

mat_ab <- subset(mat, 
                 rownames(mat) %in% grep("PABO",
	 	                                     rownames(mat), 
	 	                                     ignore.case = T, 
	 	                                     value = TRUE)
                                        )

mat_ab[1:10, 1:10]
```

```{r}
# then the rest are RNA reads
mat_rna <- subset(mat, 
	                rownames(mat) %in% grep("PABO",
		                                      rownames(mat), 
		                                      ignore.case = T, 
		                                      value = TRUE, 
		                                      invert = TRUE)
	                                       )

mat_rna[1:10, 1:10]
```

Now it's ready to pass the gene matrix to the Seurat package
```{r}
# now build Seurat object
# 1. create RNA assay
rhap <- Seurat::CreateSeuratObject(counts = mat_rna, 
		                               project = "rhap", 
		                               assay = "RNA")

# 2. Add the protein assay (if you don't have protein data, just ignore this step)
rhap[["AB"]] <- Seurat::CreateAssayObject(counts = mat_ab)

rhap
```

What if the sample were pooled before cell capture?
```{r}
# The BD Rhapsody pipeline 
# add sample tag information
rhap_tag <- "Rhapsody PBMC WTA AbSeq SMK/BD-Demo-WTA-AbSeq-SMK_Sample_Tag_Calls.csv"

tag_mat <- read.table(file = rhap_tag, 
                      header = T, 
                      sep = ",") 

tag_mat[1:5,]
```

Not all the columns are needed. Pick the one that is suitable.
```{r}
# add tag information to meta.data in rhap
rhap$tag <- tag_mat$Sample_Name

# !!! alway check if the cell indices are aligned between two datasets
identical(colnames(rhap), as.character(tag_mat$Cell_Index))

# now inspect the Seurat object rhap to see what's in there.
```

### 1b. 10X Genomics
10X Genomics system is arguably the most famous platform in scRNAseq.
Lots of functions, packages, software are built based on that.
You probably will find less hoops to jump through when trying other scRNAseq packages.

Like in here, Seurat has built a function to load 10X data.

```{r}
# loading data for 10X is simpler as Seurat has built a function to do that
# find the folder that contains the gene matrix data
tenX_file <- "10x PBMC TotalSeq WTA filtered"

# then read the files
tenX_mat <- Read10X(tenX_file)

# check the tenX object
glimpse(tenX_mat)
```

```{r}
# build the Seurat object
# 1. create RNA assay
tenX <- Seurat::CreateSeuratObject(counts = tenX_mat$`Gene Expression`, 
		                               project = "tenX", 
		                               assay = "RNA")

# 2. Add the protein assay (if you don't have protein data, just ignore this step)
tenX[["AB"]] <- Seurat::CreateAssayObject(counts = tenX_mat$`Antibody Capture`)

tenX
```

Note: 10X platform also provides results from sample demultiplexing.
This sample tag information can be added to the Seurat object the 
same way as above in the BD Rhapsody section.

### 2. Save the raw Seurat object
So that you can come back from this point and don't need to start from loading the data to R.

```{r}
# Save Seurat objects to local drive
# Create a output folder (manually)
save(rhap,
     tenX,
     file = "output/raw demo Seurat objects.RData")
```

### 3. Process the Seurat obejct

###    3a. QC check
The BD Rhapsody system allows you to view (via the cartridge scanner) what cells are captured.
So, that feature sort of gives you a hint of the quality of the cells upon capture.

On the other hand, the process of cell-bead droplet formation can't be reviewed in the 10X system.
But cells can be inspected while doing cell counting, e.g. through haemocytometer.

```{r}
# QC checks
#### Rhapsody demo data
# 1. library size and genes detected in each cell
p1 <- FeatureScatter(rhap, 
                     feature1 = "nCount_RNA", 
                     feature2 = "nFeature_RNA") + 
      ggtitle("Rhapsody")
p1

# can't tell really. display in log scale may help?
p2 <- FeatureScatter(rhap, 
                     feature1 = "nCount_RNA", 
                     feature2 = "nFeature_RNA") + 
      scale_y_log10() + 
      scale_x_log10() +
      ggtitle("Rhapsody")

p2


#### 10X demo data
p3 <- FeatureScatter(tenX, 
                     feature1 = "nCount_RNA", 
                     feature2 = "nFeature_RNA") + 
      scale_y_log10() + 
      scale_x_log10() +
      ggtitle("10X")

p3

```

If a cell had RNA leakage before capture, the ratio between mitochondrial gene (MT-**) count against 
RNA gene count in this cell is likely over the average line of all cells. Those cells with high MT% are 
usually considered as low quality cells.

```{r}
# check mitochondrial genes
# this is one of the ways to tell whether the cells are intact.
rhap <- Seurat::PercentageFeatureSet(rhap, 
                                     pattern = "^MT[.-]", 
                                     col.name = "percent.mt")

tenX <- Seurat::PercentageFeatureSet(tenX, 
                                     pattern = "^MT-", 
                                     col.name = "percent.mt")

# the MT percentages on a violin plot
p4 <- VlnPlot(rhap, features = "percent.mt")
p5 <- VlnPlot(tenX, features = "percent.mt")

wrap_plots(p4 + p5)
p4
p5
```
```{r}
# remove poor quality cells
#### for the Rhapsody data set
# nFeature_RNA < 200 & percent.mt > 25
rhap <- subset(rhap, 
               subset = percent.mt < 25 & 
               nFeature_RNA > 200, 
               invert = F)

#### for 10x data set
# nFeature_RNA < 200 & percent.mt > 25
tenX <- subset(tenX, 
               subset = percent.mt < 25 & 
               nFeature_RNA > 200, 
               invert = F)

p6 <- FeatureScatter(rhap, 
                     feature1 = "nCount_RNA", 
                     feature2 = "nFeature_RNA") + 
      scale_y_log10() + 
      scale_x_log10() +
      ggtitle("Rhapsody")

p7 <- FeatureScatter(tenX, 
                     feature1 = "nCount_RNA", 
                     feature2 = "nFeature_RNA") + 
      scale_y_log10() + 
      scale_x_log10() +
      ggtitle("10X")

p8 <- VlnPlot(rhap, features = "percent.mt")
p9 <- VlnPlot(tenX, features = "percent.mt")

p6 + p7 + p8 + p9 
```

### 3b. Normalization and scaling
The purpose of normalization:
1. library sizes can be different among cells
2. efficiency in cDNA capture or PCR amplification can be different 
(read more in here: https://bioconductor.org/books/3.15/OSCA.basic/normalization.html#normalization-transformation)
3. gene size varies

The purpose of scaling:
1. find the highly variable genes
2. use both average gene expression and dispersion to calculate the z-score
(read more in here: https://learn.gencore.bio.nyu.edu/single-cell-rnaseq/seurat-part-3-data-normalization/)
3. scaled data will be fed into PCA calculation

```{r}
# Normalize the data
# Default Seurat method
# LogNormalize: Feature counts for each cell are divided by the total counts 
# for that cell and multiplied by the scale.factor. 
# This is then natural-log transformed using log1p.

# This step is to remove biases from sequencing depth, library size or 
# PCR amplification difference among cells

# Rhap Seurat
# Process protein data
Seurat::DefaultAssay(rhap) <- 'AB'
  
# Normalize and scale data
rhap <- rhap %>% 
        Seurat::NormalizeData(normalization.method = 'CLR', 
                              margin = 2)  %>% 
        Seurat::FindVariableFeatures() %>% 
        Seurat::ScaleData() 

# perform PCA
rhap <- Seurat::RunPCA(object = rhap, 
                       reduction.name = 'apca')

# perform UMAP
rhap <- Seurat::RunUMAP(rhap, 
                        reduction = 'apca', 
                        dims = 1:7, 
                        assay = 'AB', 
                        reduction.name = 'adt.umap',
			                  reduction.key = 'adtUMAP_')

# Find clusters
rhap <- Seurat::FindNeighbors(rhap, 
                              reduction = "apca", 
                              dims = 1:7)

rhap <- Seurat::FindClusters(rhap, 
                             resolution = 0.8, 
                             graph.name = "AB_snn")

# check clusters by protein data
DimPlot(rhap, 
        reduction = "adt.umap", 
        group.by = "seurat_clusters")

# Process RNA data
Seurat::DefaultAssay(rhap) <- 'RNA'

# find top most variant genes
rhap <- rhap %>% 
        Seurat::NormalizeData() %>% 
		    Seurat::FindVariableFeatures(., selection.method = "vst")

# Identify the 10 most highly variable genes
top20 <- head(VariableFeatures(rhap), 20)

# check highly variable genes
p10 <- VariableFeaturePlot(rhap)
p11 <- LabelPoints(plot = p10, points = top20, repel = TRUE)
p11

# scale data
rhap <- Seurat::ScaleData(rhap, 
                          verbose = FALSE)

# perform PCA
rhap <- Seurat::RunPCA(rhap,
                       verbose = FALSE)

# Elbow plot
# Inspect top X numbers of principle components to check the degree of variance
# in each component.
ElbowPlot(rhap, 
          ndims = 50)

# perform UMAP
rhap <- Seurat::RunUMAP(rhap, 
                        reduction = "pca", 
                        dims = 1:30)

# Find clusters
rhap <- Seurat::FindNeighbors(rhap,
                              reduction = "pca", 
                              dims = 1:30)

rhap <- Seurat::FindClusters(rhap, 
                             resolution = 0.8)

rhap <- Seurat::BuildClusterTree(rhap)

# check clusters by RNA content
DimPlot(rhap, 
        reduction = "umap")
```

```{r}
# 10X Seurat
# Rhap Seurat
# Process protein data
Seurat::DefaultAssay(tenX) <- 'AB'
  
# Normalize and scale data
tenX <- tenX %>% 
        Seurat::NormalizeData(normalization.method = 'CLR', 
                              margin = 2)  %>% 
        Seurat::FindVariableFeatures() %>% 
        Seurat::ScaleData() 

# perform PCA
tenX <- Seurat::RunPCA(object = tenX, 
                       reduction.name = 'apca')

# perform UMAP
tenX <- Seurat::RunUMAP(tenX, 
                        reduction = 'apca', 
                        dims = 1:7, 
                        assay = 'AB', 
                        reduction.name = 'adt.umap',
			                  reduction.key = 'adtUMAP_')

# Find clusters
tenX <- Seurat::FindNeighbors(tenX, 
                              reduction = "apca", 
                              dims = 1:7)

tenX <- Seurat::FindClusters(tenX, 
                             resolution = 0.8, 
                             graph.name = "AB_snn")

# check clusters by protein data
DimPlot(tenX, 
        reduction = "adt.umap", 
        group.by = "seurat_clusters")

# Process RNA data
Seurat::DefaultAssay(tenX) <- 'RNA'

# find top most variant genes
tenX <- tenX %>% 
        Seurat::NormalizeData() %>% 
		    Seurat::FindVariableFeatures(., selection.method = "vst")

# Identify the 10 most highly variable genes
top20 <- head(VariableFeatures(tenX), 20)

# check highly variable genes
p10 <- VariableFeaturePlot(tenX)
p11 <- LabelPoints(plot = p10, points = top20, repel = TRUE)
p11

# scale data
tenX <- Seurat::ScaleData(tenX, 
                          verbose = FALSE)

# perform PCA
tenX <- Seurat::RunPCA(tenX,
                       verbose = FALSE)

# Elbow plot
# Inspect top X numbers of principle components to check the degree of variance
# in each component.
ElbowPlot(tenX, 
          ndims = 50)

# perform UMAP
tenX <- Seurat::RunUMAP(tenX, 
                        reduction = "pca", 
                        dims = 1:30)

# Find clusters
tenX <- Seurat::FindNeighbors(tenX,
                              reduction = "pca", 
                              dims = 1:30)

tenX <- Seurat::FindClusters(tenX, 
                             resolution = 0.8)

tenX <- Seurat::BuildClusterTree(tenX)

# check clusters by RNA content
DimPlot(tenX, 
        reduction = "umap")
```


### 3c. Play with UMAP and cluster parameters
```{r}
# umap
# use ?RunUMAP to view the full description of each parameter.
# Selection here is mainly for global & local structures of the embedded points.
rhap <- RunUMAP(
                rhap,
                dims = 1:30,
                umap.method = "uwot",    # other methods: uwot-learn, umap-learn
                n.neighbors = 30L,       # 5-50, larger number to preserve global structure
                min.dist = 0.3,          # 0.001 - 0.5 how tightly the points to be. Higher values to 
                                         # get more evenly distribution.
                spread = 1,              # the degree of points spreading out
                seed.use = 42,
        )

DimPlot(rhap, 
        reduction = "umap") + ggtitle("default setitngs - UMAP")

rhap <- RunUMAP(
                rhap,
                dims = 1:30,
                umap.method = "uwot",    # other methods: uwot-learn, umap-learn
                n.neighbors = 40L,       # 5-50, larger number to preserve global structure
                min.dist = 0.5,          # 0.001 - 0.5 how tightly the points to be. Higher values to 
                                         # get more evenly distribution.
                spread = 3,              # the degree of points spreading out
                seed.use = 42,
        )

DimPlot(rhap, 
        reduction = "umap") + ggtitle("More global structure - UMAP")

# cluster
# Two steps:
# 1. Calculate the nearest-neighbor graph
rhap <- FindNeighbors(rhap,
                      dims = 1:30) # PCA dimensions to use

# 2. Assign cells to clusters based on the KNN information above
rhap <- FindClusters(rhap,
                     resolution = 0.5) # larger value for more communities (clusters)

DimPlot(rhap, 
        reduction = "umap") + ggtitle("res 0.5 - clustering")

# let's change the resolution value and see how it looks.
rhap <- FindClusters(rhap,
                     resolution = 0.05) # larger value for more communities (clusters)

DimPlot(rhap, 
        reduction = "umap") + ggtitle("res 0.05 - clustering")

rhap <- FindClusters(rhap,
                     resolution = 1) # larger value for more communities (clusters)

DimPlot(rhap, 
        reduction = "umap") + ggtitle("res 1 - clustering")

rhap <- FindClusters(rhap,
                     resolution = 5) # larger value for more communities (clusters)

DimPlot(rhap, 
        reduction = "umap") + ggtitle("res 5 - clustering")

# resolution = 0.5 seems good. Let's use that.
rhap <- FindClusters(rhap,
                     resolution = 0.5) # larger value for more communities (clusters)

```

### 3d. Differential gene expression
Now you have cells in different groups. 
Let's find out what genes are up-regulated in each group against the rest.

```{r}
# find marker genes
# get more inforamtion about this FindAllMarkers function by running this line in the console: ?FindAllMarker
rhap.markers <- FindAllMarkers(rhap, 
                               only.pos = TRUE, # only calculate the up-regulated genes. 
                                                # Set it to FALSE to get down-regulated genes
                               )

# Inspect the marker gene matrix
# Subset the top 5 marker genes of each group
rhap.markers.short <- rhap.markers %>%
                      group_by(cluster) %>%
                      slice_max(n = 5, 
                      order_by = avg_log2FC)

head(rhap.markers.short)
```

## Save data
```{r}
# Save Seurat objects to local drive
save(rhap,
     tenX,
     rhap.markers,
     file = "output/Seurat with DGE.RData")
```

## Back to PowerPoint talk for advanced analysis











